### PR TITLE
[RFC] Deprecate ReST API

### PR DESCRIPTION
**Pull request by jordisala1991**
## Subject

I am targeting this branch, because this is BC.

## Changelog

```markdown
### Deprecated
- Deprecated ReST API with FOSRest, Nelmio Api Docs and JMS Serializer.
```

### CODE REVIEW

**Comment by jordisala1991** [comment]
With this deprecation there are many things that are solved.

1. We stop offering an API that becomes outdated very quickly since each version of Nelmio changes completely the annotations
2. We can remove a lot of dependencies: fosRest, nelmio, JMS serializer
3. We don't have to worry about Datagrid-bundle anymore because on 4.0 it won't be needed.

I think we should take this PR as the way to go for the rest of bundles and apply the same. wdyt @sonata-project/contributors ?

File: UPGRADE-3.x.md [hunk 1]
```diff
@@ -4,6 +4,12 @@ UPGRADE 3.x
 UPGRADE FROM 3.x to 3.x
 =======================
 
+### Deprecate API
+
+Integration with FOSRest, JMS Serializer and Nelmio Api Docs is deprecated, the API provided with this bundle will be removed on 4.0.
```
Comment (phansys) [inline UPGRADE-3.x.md:9]:
```suggestion
Integration with FOSRest, JMS Serializer and Nelmio Api Docs is deprecated, the ReST API provided with this bundle will be removed on 4.0.
```

File: UPGRADE-3.x.md [hunk 2]
```diff
@@ -4,6 +4,12 @@ UPGRADE 3.x
...
+
+If you are relying on this, consider moving to other solution like Api Platform instead.
```
Comment (phansys) [inline UPGRADE-3.x.md:11]:
```suggestion
If you are relying on this, consider moving to other solution like [API Platform](https://api-platform.com/) instead.
```

**Review by phansys — CHANGES_REQUESTED** [review]
IMO, the title and the changelog must also reflect that we are talking specifically about the ReST API.

**Review by phansys — DISMISSED** [review]

**Review by VincentLanglet — COMMENTED** [review]
Does someone in the @sonata-project/contributors is using the Rest Api ?

File: src/Extra/ApiMediaFile.php [hunk 3]
```diff
@@ -16,7 +16,11 @@
 use Symfony\Component\HttpFoundation\File\File;
 
 /**
+ * NEXT_MAJOR: Remove this file.
+ *
  * @final since sonata-project/media-bundle 3.21.0
```
Comment (core23) [inline src/Extra/ApiMediaFile.php:21]: IMHO you can remove the `@final` information

**Review by core23 — CHANGES_REQUESTED** [review]

File: src/Extra/ApiMediaFile.php (hunk 3 above)
Comment (jordisala1991) [reply src/Extra/ApiMediaFile.php:21]: Does it matter? I will remove it once I do the manual merge

Comment (core23) [reply src/Extra/ApiMediaFile.php:21]: Yes, because this information is for the user and can be very confusing. It could either be final or deprecated.

Comment (jordisala1991) [reply src/Extra/ApiMediaFile.php]: I don't see how final since 3.21 and deprecated since 3.34 could confuse our users, but changed!.

**Review by wbloszyk — APPROVED** [review]

**Review by core23 — APPROVED** [review]

**Review by VincentLanglet — APPROVED** [review]

**Comment by jordisala1991** [comment]
3 approvals + @phansys did an approval before. Lets do this.

**Pull request merged**